File: lib/ios/VesperPlayerKit/Sources/VesperPlayerKit/Bridge/PlayerBridgeFactory.swift
import Foundation

@MainActor
enum PlayerBridgeFactory {
    private static let defaultBackend: PlayerBridgeBackend = .rustNativeStub

    static func defaultBridgeBackend() -> PlayerBridgeBackend {
        defaultBackend
    }

    static func makeDefaultBridge(
        initialSource: VesperPlayerSource? = nil,
        resiliencePolicy: VesperPlaybackResiliencePolicy = VesperPlaybackResiliencePolicy(),
        trackPreferencePolicy: VesperTrackPreferencePolicy = VesperTrackPreferencePolicy(),
        preloadBudgetPolicy: VesperPreloadBudgetPolicy = VesperPreloadBudgetPolicy(),
        keepScreenOnDuringPlayback: Bool = true,
        benchmarkConfiguration: VesperBenchmarkConfiguration = .disabled,
        sourceNormalizerConfiguration: VesperSourceNormalizerConfiguration =
            VesperSourceNormalizerConfiguration(),
        frameProcessorConfiguration: VesperFrameProcessorConfiguration =
            VesperFrameProcessorConfiguration(),
        nativeFramePipelineConfiguration: VesperNativeFramePipelineConfiguration =
            VesperNativeFramePipelineConfiguration(),
        pipelineEventHookConfiguration: VesperPipelineEventHookConfiguration =
            VesperPipelineEventHookConfiguration()
    ) -> VesperPlayerController {
        switch defaultBackend {
        case .fakeDemo:
            return VesperPlayerController(
                FakePlayerBridge(
                    initialSource: initialSource,
                    resiliencePolicy: resiliencePolicy,
                    trackPreferencePolicy: trackPreferencePolicy,
                    preloadBudgetPolicy: preloadBudgetPolicy,
                    benchmarkConfiguration: benchmarkConfiguration
                ),
                keepScreenOnDuringPlayback: keepScreenOnDuringPlayback
            )
        case .rustNativeStub:
            return VesperPlayerController(
                VesperNativePlayerBridge(
                    initialSource: initialSource,
                    resiliencePolicy: resiliencePolicy,
                    trackPreferencePolicy: trackPreferencePolicy,
                    preloadBudgetPolicy: preloadBudgetPolicy,
                    benchmarkConfiguration: benchmarkConfiguration,
                    sourceNormalizerConfiguration: sourceNormalizerConfiguration,
                    frameProcessorConfiguration: frameProcessorConfiguration,
                    nativeFramePipelineConfiguration: nativeFramePipelineConfiguration,
                    pipelineEventHookConfiguration: pipelineEventHookConfiguration
                ),
                keepScreenOnDuringPlayback: keepScreenOnDuringPlayback
            )
        }
    }
}

@MainActor
public enum VesperPlayerControllerFactory {
    public static func defaultBackend() -> PlayerBridgeBackend {
        PlayerBridgeFactory.defaultBridgeBackend()
    }

    public static func makeDefault(
        initialSource: VesperPlayerSource? = nil,
        resiliencePolicy: VesperPlaybackResiliencePolicy = VesperPlaybackResiliencePolicy(),
        trackPreferencePolicy: VesperTrackPreferencePolicy = VesperTrackPreferencePolicy(),
        preloadBudgetPolicy: VesperPreloadBudgetPolicy = VesperPreloadBudgetPolicy(),
        keepScreenOnDuringPlayback: Bool = true,
        benchmarkConfiguration: VesperBenchmarkConfiguration = .disabled,
        sourceNormalizerConfiguration: VesperSourceNormalizerConfiguration =
            VesperSourceNormalizerConfiguration(),
        frameProcessorConfiguration: VesperFrameProcessorConfiguration =
            VesperFrameProcessorConfiguration(),
        nativeFramePipelineConfiguration: VesperNativeFramePipelineConfiguration =
            VesperNativeFramePipelineConfiguration(),
        pipelineEventHookConfiguration: VesperPipelineEventHookConfiguration =
            VesperPipelineEventHookConfiguration()
    ) -> VesperPlayerController {
        PlayerBridgeFactory.makeDefaultBridge(
            initialSource: initialSource,
            resiliencePolicy: resiliencePolicy,
            trackPreferencePolicy: trackPreferencePolicy,
            preloadBudgetPolicy: preloadBudgetPolicy,
            keepScreenOnDuringPlayback: keepScreenOnDuringPlayback,
            benchmarkConfiguration: benchmarkConfiguration,
            sourceNormalizerConfiguration: sourceNormalizerConfiguration,
            frameProcessorConfiguration: frameProcessorConfiguration,
            nativeFramePipelineConfiguration: nativeFramePipelineConfiguration,
            pipelineEventHookConfiguration: pipelineEventHookConfiguration
        )
    }

    public static func probePlaybackCapability(
        _ request: VesperPlaybackCapabilityProbeRequest
    ) -> VesperPlaybackCapabilityProbeResult {
        VesperPlaybackCapabilityProbe.probe(
            VesperPlaybackCapabilityProbeRequest(
                source: request.source,
                codec: request.codec,
                width: request.width,
                height: request.height,
                frameRate: request.frameRate,
                requiresNativeFrame: request.requiresNativeFrame,
                sourceNormalizerConfiguration: request.sourceNormalizerConfiguration,
                frameProcessorConfiguration: request.frameProcessorConfiguration,
                nativeFramePipelineConfiguration: request.nativeFramePipelineConfiguration
            ),
            sessionProbeProvider: VesperIOSSessionProbeProvider.currentDisplay()
        )
    }
}

File: lib/android/vesper-player-kit/src/main/java/io/github/ikaros/vesper/player/android/bridge/PlayerBridgeFactory.kt
package io.github.ikaros.vesper.player.android

import android.content.Context

internal object PlayerBridgeFactory {
    private val defaultBackend = PlayerBridgeBackend.VesperNativeStub

    fun createDefault(
        context: Context,
        initialSource: VesperPlayerSource? = null,
        resiliencePolicy: VesperPlaybackResiliencePolicy = VesperPlaybackResiliencePolicy(),
        trackPreferencePolicy: VesperTrackPreferencePolicy = VesperTrackPreferencePolicy(),
        preloadBudgetPolicy: VesperPreloadBudgetPolicy = VesperPreloadBudgetPolicy(),
        decoderBackend: VesperDecoderBackend = VesperDecoderBackend.SystemOnly,
        surfaceKind: NativeVideoSurfaceKind = NativeVideoSurfaceKind.SurfaceView,
        keepScreenOnDuringPlayback: Boolean = true,
        benchmarkConfiguration: VesperBenchmarkConfiguration = VesperBenchmarkConfiguration.Disabled,
        sourceNormalizerConfiguration: VesperSourceNormalizerConfiguration =
            VesperSourceNormalizerConfiguration(),
        frameProcessorConfiguration: VesperFrameProcessorConfiguration =
            VesperFrameProcessorConfiguration(),
        nativeFramePipelineConfiguration: VesperNativeFramePipelineConfiguration =
            VesperNativeFramePipelineConfiguration(),
        pipelineEventHookConfiguration: VesperPipelineEventHookConfiguration =
            VesperPipelineEventHookConfiguration(),
    ): PlayerBridge =
        when (defaultBackend) {
            PlayerBridgeBackend.FakeDemo -> {
                require(pipelineEventHookConfiguration.pluginReferences.isEmpty()) {
                    "Fake demo players do not support Android playback event-hook references"
                }
                FakePlayerBridge(
                    initialSource = initialSource,
                    resiliencePolicy = resiliencePolicy,
                    trackPreferencePolicy = trackPreferencePolicy,
                    preloadBudgetPolicy = preloadBudgetPolicy,
                    keepScreenOnDuringPlayback = keepScreenOnDuringPlayback,
                    benchmarkConfiguration = benchmarkConfiguration,
                    appContext = context.applicationContext,
                )
            }
            PlayerBridgeBackend.VesperNativeStub -> {
                val appContext = context.applicationContext
                val resolvedPluginArtifacts =
                    VesperBundledPluginResolver.resolve(
                        context = appContext,
                        sourceNormalizerConfiguration = sourceNormalizerConfiguration,
                        frameProcessorConfiguration = frameProcessorConfiguration,
                        nativeFramePipelineConfiguration = nativeFramePipelineConfiguration,
                    )
                val pipelineEventHookRegistryOwner =
                    pipelineEventHookConfiguration.pluginReferences
                        .takeIf { it.isNotEmpty() }
                        ?.let { references ->
                            DefaultVesperPluginRegistryFactory.create(appContext, references)
                        }
                try {
                    val benchmarkRecorder =
                        VesperBenchmarkRecorder(
                            configuration = benchmarkConfiguration,
                            context = appContext,
                        )
                    VesperNativePlayerBridge(
                        bindings =
                            VesperNativeJniBindings(
                                context = appContext,
                                preloadBudgetPolicy = preloadBudgetPolicy,
                                decoderBackend = decoderBackend,
                                benchmarkRecorder = benchmarkRecorder,
                                sourceNormalizerConfiguration = sourceNormalizerConfiguration,
                                resolvedPluginArtifacts = resolvedPluginArtifacts,
                                pipelineEventHookRegistryHandle =
                                    pipelineEventHookRegistryOwner?.handle ?: 0L,
                                pipelineEventHookReferencesJson =
                                    encodeVesperPluginReferences(
                                        pipelineEventHookConfiguration.pluginReferences,
                                    ),
                            ),
                        initialSource = initialSource,
                        currentResiliencePolicy = resiliencePolicy,
                        trackPreferencePolicy = trackPreferencePolicy,
                        preloadBudgetPolicy = preloadBudgetPolicy,
                        decoderBackend = decoderBackend,
                        benchmarkRecorder = benchmarkRecorder,
                        keepScreenOnDuringPlayback = keepScreenOnDuringPlayback,
                        appContext = appContext,
                        surfaceKind = surfaceKind,
                        sourceNormalizerConfiguration = sourceNormalizerConfiguration,
                        frameProcessorConfiguration = frameProcessorConfiguration,
                        nativeFramePipelineConfiguration = nativeFramePipelineConfiguration,
                        pipelineEventHookRegistryOwner = pipelineEventHookRegistryOwner,
                    )
                } catch (error: Throwable) {
                    pipelineEventHookRegistryOwner?.close()
                    throw error
                }
            }
        }
}
